### DISCUSSION TITLE
"Human in a loop" / Skyvern as Copilot / "AI Driven iterative co-creation of workflows based on automatic collection of `user stories` traces"

### DISCUSSION THREAD

**Discussion by gwpl** [Skyvern-AI/skyvern#1420] [Ideas]
# "Human in a loop" / Skyvern as Copilot / "AI Driven iterative co-creation of workflows based on automatic collection of `user stories` traces"

I would love to use skyvern in copilot way!

I.e. to be able to switch between "me driving it" and "skyvern driving it".

The other question would be probably how to make it work not only in headfull but also headless mode, in browsertab in skyvern cloud setting, but imagine this:

often in given workflows only some parts are tedious but simple and AI can help with easy prompting,
and other parts maybe so expensive to prompt/workaround to force ai to "fully autopilot" handle, that maybe easier, if system woudl pause and do "human in the loop" "copilot" style collaboration...

Now the best part, ability to opt-in to record details of this joined interaction and human actions , could allow to have traces that could be used to generate automatically suggestions which parts of workflow done by human can be automated,

allowing continous development from "mostly manual" to "more and more refined and automated" until "fully automated",
and if at any point one of the parts of workflow would start breaking (e.g. website change, requirements change, etc),

it could be first replaced with "human-in-a-loop" step, and then again continously refined with suggestions until fully automated again,

never breaking actual business workflows continuity! (for businesses, I am not talking latency critical SaaS offerings, but businesses that could be supported by skyvern)

**Comment by suchintan** [comment]
This is A good idea and something we've thought about building

**Comment by aaronlab** [comment]
This pattern seems strongest if the trace records handoff boundaries, not only final workflow state.

For a human-in-the-loop / copilot mode, the useful minimum trace shape might be:

- actor: human or Skyvern;
- intent for the current step;
- URL/title or route pattern;
- selected action and target when one exists;
- reason for handoff: uncertainty, auth, cost, site change, missing selector, policy, etc.;
- result: completed, skipped, failed, or converted into a reusable workflow step;
- optional redacted visual/state artifact for review.

That gives the workflow a safe middle state: the business process keeps running manually where needed, while the recorded trace creates evidence for which handoffs are stable enough to convert into automation later.

For BrowserTrace I have been working on the narrower debugging version of this idea: local traces that preserve browser-agent step provenance, status, errors, screenshots/URLs when safe, and redacted shareable exports. The same boundary seems useful here, but with an explicit `actor` and `handoff_reason` field so the trace can support progressive automation rather than only post-failure debugging.

If Skyvern ever exposes hooks around human action, agent action, pause/resume, and workflow step state, that would probably be enough to prototype this without requiring a full DOM snapshot or full session recording as the first version.
